include/albatross/src/eigen/serializable_ldlt.hpp

```diff
@@ -52,20 +53,66 @@ class SerializableLDLT : public LDLT<MatrixXd, Lower> {
 
   bool is_initialized() const { return this->m_isInitialized; }
 
+  /*
+   * Computes the inverse of the square root of the diagonal, D^{-1/2}
+   */
+  Eigen::DiagonalMatrix<double, Eigen::Dynamic> diagonal_sqrt_inverse() const {
+    Eigen::VectorXd thresholded_diag_sqrt_inverse(this->vectorD());
+    for (Eigen::Index i = 0; i < thresholded_diag_sqrt_inverse.size(); ++i) {
+      if (thresholded_diag_sqrt_inverse[i] > 0.) {
+        thresholded_diag_sqrt_inverse[i] =
+            1. / std::sqrt(thresholded_diag_sqrt_inverse[i]);
+      } else {
+        thresholded_diag_sqrt_inverse[i] = 0.;
+      }
+    }
+    return thresholded_diag_sqrt_inverse.asDiagonal();
+  }
+
+  /*
+   * Computes the square root of the diagonal, D^{1/2}
+   */
+  Eigen::DiagonalMatrix<double, Eigen::Dynamic> diagonal_sqrt() const {
+    Eigen::VectorXd thresholded_diag = this->vectorD();
+    for (Eigen::Index i = 0; i < thresholded_diag.size(); ++i) {
+      if (thresholded_diag[i] > 0.) {
+        thresholded_diag[i] = std::sqrt(thresholded_diag[i]);
+      } else {
+        thresholded_diag[i] = 0.;
```

Comment:
Should these functions be asserting the diagonal is nonnegative? Can we get a zero diagonal? It's valid for this function, but a problem everywhere else in LDLT.

Reply:
While in theory we should never get a negative diagonal for a PSD matrix numerically we can and this logic amounts to zeroing out any of the numerically negative definitely directions and mimics what Eigen does internally:

https://github.com/libigl/eigen/blob/master/Eigen/src/Cholesky/LDLT.h#L578-L584

We actually assert that the diagonal is positive here:

https://github.com/swift-nav/albatross/blob/master/include/albatross/src/eigen/serializable_ldlt.hpp#L39

but I'm not convinced that's the correct thing to do, we should probably just be warning and continuing but unfortunately we don't have logging setup in albatross yet.